library/gui/screen/src/main/java/org/quiltmc/qsl/screen/mixin/client/ScreenMixin.java

```diff
@@ -0,0 +1,289 @@
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import net.minecraft.client.gui.Selectable;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.font.TextRenderer;
+import net.minecraft.client.gui.Drawable;
+import net.minecraft.client.gui.Element;
+import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.gui.widget.ClickableWidget;
+import net.minecraft.client.render.item.ItemRenderer;
+
+@Mixin(Screen.class)
+abstract class ScreenMixin implements ScreenExtensions, QuiltScreenHooks {
+	@Shadow
+	@Final
+	private List<Selectable> selectables;
+	@Shadow
+	@Final
+	private List<Element> children;
+	@Shadow
+	@Final
+	private List<Drawable> drawables;
+
+	@Shadow
+	private MinecraftClient client;
+
+	@Shadow
+	private ItemRenderer itemRenderer;
+
+	@Shadow
+	private TextRenderer textRenderer;
+
+	@Unique
+	private ButtonList quilt$quiltButtons;
+	@Unique
+	private Event<ScreenEvents.Remove> quilt$removeEvent;
+	@Unique
+	private Event<ScreenEvents.BeforeTick> quilt$beforeTickEvent;
+	@Unique
+	private Event<ScreenEvents.AfterTick> quilt$afterTickEvent;
+	@Unique
+	private Event<ScreenEvents.BeforeRender> quilt$beforeRenderEvent;
+	@Unique
+	private Event<ScreenEvents.AfterRender> quilt$afterRenderEvent;
+
+	// Keyboard
+	@Unique
+	private Event<ScreenKeyboardEvents.AllowKeyPress> quilt$allowKeyPressEvent;
+	@Unique
+	private Event<ScreenKeyboardEvents.BeforeKeyPress> quilt$beforeKeyPressEvent;
+	@Unique
+	private Event<ScreenKeyboardEvents.AfterKeyPress> quilt$afterKeyPressEvent;
+	@Unique
+	private Event<ScreenKeyboardEvents.AllowKeyRelease> quilt$allowKeyReleaseEvent;
+	@Unique
+	private Event<ScreenKeyboardEvents.BeforeKeyRelease> quilt$beforeKeyReleaseEvent;
+	@Unique
+	private Event<ScreenKeyboardEvents.AfterKeyRelease> quilt$afterKeyReleaseEvent;
+
+	// Mouse
+	@Unique
+	private Event<ScreenMouseEvents.AllowMouseClick> quilt$allowMouseClickEvent;
+	@Unique
+	private Event<ScreenMouseEvents.BeforeMouseClick> quilt$beforeMouseClickEvent;
+	@Unique
+	private Event<ScreenMouseEvents.AfterMouseClick> quilt$afterMouseClickEvent;
+	@Unique
+	private Event<ScreenMouseEvents.AllowMouseRelease> quilt$allowMouseReleaseEvent;
+	@Unique
+	private Event<ScreenMouseEvents.BeforeMouseRelease> quilt$beforeMouseReleaseEvent;
+	@Unique
+	private Event<ScreenMouseEvents.AfterMouseRelease> quilt$afterMouseReleaseEvent;
+	@Unique
+	private Event<ScreenMouseEvents.AllowMouseScroll> quilt$allowMouseScrollEvent;
+	@Unique
+	private Event<ScreenMouseEvents.BeforeMouseScroll> quilt$beforeMouseScrollEvent;
+	@Unique
+	private Event<ScreenMouseEvents.AfterMouseScroll> quilt$afterMouseScrollEvent;
+
+	@Inject(method = "init(Lnet/minecraft/client/MinecraftClient;II)V", at = @At("HEAD"))
+	private void beforeInitScreen(MinecraftClient client, int width, int height, CallbackInfo ci) {
```

Comment:
This is a lot of object creation. Why does each screen need to create these events? Is there a reason a global event with a screen argument cannot be used? Users of the event can simply use an equality check on the screen object.

Reply:
Doesn't seem to be pointed out in the original FAPI Screen API PR.

Reply:
> This is a lot of object creation. Why does each screen need to create these events? Is there a reason a global event with a screen argument cannot be used? Users of the event can simply use an equality check on the screen object.

Because you can do things like "ScreenEvents.afterRender(screen).register". Technically i can just make these global events. Then I can make a quick and simple wrapper function to allow for that same behavior.

Reply:
It may be easier but it makes listening to global events (all screens) impossible and is less efficient.

Reply:
Ok. i will start working on this

Reply:
Forgot to say this earlier but fixed.